iroha/src/peer.rs

```diff
@@ -18,6 +22,247 @@ pub mod isi {
     /// synchronization components will start using it.
     #[derive(Clone, Debug, PartialEq, Io, IntoContract, Encode, Decode)]
     pub struct AddPeer {
-        pub peer: Peer,
+        pub address: String,
+        pub peer_key: PublicKey,
+    }
+}
+
+const PING_SIZE: usize = 32;
+
+#[derive(Io, Decode, Encode, Debug, Clone)]
+enum Message {
+    //TODO: introduce other features like block sync, voting and etc.
+    Ping(Ping),
+    Pong(Ping),
+    PendingTx(Transaction),
+    AddPeer(PeerId),
+    NewPeer(PeerId),
+    RemovePeer(PeerId),
+}
+
+#[derive(Encode, Decode, PartialEq, Eq, Debug, Clone, Hash)]
+struct Ping {
+    payload: Vec<u8>,
+    to_peer: PeerId,
+    from_peer: PeerId,
+}
+
+#[derive(Encode, Decode, PartialEq, Eq, Debug, Clone, Hash)]
+struct PeerId {
+    listen_address: String,
+}
+
+struct NodeState {
+    pub peers: HashSet<PeerId>,
+    pub sent_pings: HashMap<Ping, Duration>,
+    pub listen_address: String,
+}
+
+pub struct Peer {
+    state: State<NodeState>,
+    ping_interval_sec: usize,
+    tx_interval_sec: usize,
+}
+
+#[allow(unused)]
+impl Peer {
+    pub fn new(listen_address: String, tx_interval_sec: usize, ping_interval_sec: usize) -> Peer {
+        Peer {
+            state: Arc::new(Mutex::new(NodeState {
+                peers: HashSet::new(),
+                sent_pings: HashMap::new(),
+                listen_address,
+            })),
+            ping_interval_sec,
+            tx_interval_sec,
+        }
+    }
+
+    pub async fn start(&self) -> Result<(), String> {
+        let listen_future = self.listen_and_reconnect().fuse();
+        let tx_future = self.start_broadcasting_tx().fuse();
+        let ping_future = self.start_ping().fuse();
+
```

Comment:
please do not use empty lines inside method bodies

Reply:
@humb1t Is it a general guideline? If so maybe we can add it to `CONTRIBUTING.rst`. I usually use empty lines to separate logical blocks and in general improve readability.

Reply:
Yes, will add